onert-micro/eval-driver/TrainingDriver.cpp

```diff
@@ -100,118 +93,264 @@ int entry(int argc, char **argv)
 
-  settings.metric = luci_interpreter::training::MAE;
-  onert_micro_training.test(num_train_data_samples, reinterpret_cast<const uint8_t *>(train_data),
-                            reinterpret_cast<const uint8_t *>(label_train_data),
-                            reinterpret_cast<void *>(&mae_result));
+  // Test data
+  std::unique_ptr<float []> test_input_data(new float [input_size]);
+  std::unique_ptr<float []> test_target_data(new float [target_size]);
 
-  printf("MSE_ERROR TRAIN = %f\n", mse_result);
+  for (uint32_t e = 0; e < training_epochs; ++e)
+  {
+    train_interpreter.set_training_mode(true);
+    std::cout << "Run training for epoch: " << e + 1 << "/" << training_epochs << "\n";
+    std::vector<int> predicted_labels;
+    std::vector<int> targets_labels;
+    for (int i = 0; i < num_train_data_samples; i += config.train_config.batches)
+    {
+      for (int batch = 0; batch < config.train_config.batches and i + batch < num_train_data_samples; ++batch)
+      {
+        readDataFromFile(input_input_train_data_path, reinterpret_cast<char *>(train_input_data.get()),
+                         sizeof(MODEL_TYPE) * input_size, sizeof(MODEL_TYPE) * input_size * (i + batch));
+
+        readDataFromFile(input_target_train_data_path, reinterpret_cast<char *>(train_target_data.get()),
+                         sizeof(MODEL_TYPE) * target_size, sizeof(MODEL_TYPE) * target_size * (i + batch));
+
+#if 0
+       printDataVector(train_input_data.get(), input_size, i + batch);
+#endif
+
+        train_interpreter.allocateInputs();
+        // Copy input data
+        auto *cur_train_data = train_input_data.get();
+        auto cur_input_data = train_interpreter.getInputDataAt(0);
+        std::memcpy(cur_input_data, cur_train_data, sizeof(MODEL_TYPE) * input_size);
+
+        train_interpreter.forward();
+        train_interpreter.allocateTargets();
+
+        // Copy targets values
+        auto *cur_train_target_data = train_target_data.get();
+        auto cur_target_data = train_interpreter.getTargetDataAt(0);
+        std::memcpy(cur_target_data, cur_train_target_data,
+                    sizeof(MODEL_TYPE) * target_size);
+
+        predicted_labels.push_back(predicted_class(reinterpret_cast<float *>(train_interpreter.getOutputDataAt(0)), target_size));
+        targets_labels.push_back(predicted_class(reinterpret_cast<float *>(cur_train_target_data), target_size));
+
+#if 0
+        calculateCrossEntropy(reinterpret_cast<float *>(train_interpreter.getOutputDataAt(0)),
+                              reinterpret_cast<float *>(cur_train_target_data),
+                              target_size);
+        printPredAndTargetsValues(reinterpret_cast<float *>(train_interpreter.getOutputDataAt(0)),
+                                  reinterpret_cast<float *>(cur_train_target_data),
+                                  target_size);
+#endif
+        train_interpreter.backward();
```

Comment:
@BalyshevArtem This approach (backward for each sample in a batch) does not produce the same result with backward mean/sum over batch approach ?

Reply:
I'm not sure I understood the question correctly. In this approach, for each current data example, we calculate the gradient and summarize it for all examples in the current batch sample. And then, using the `updateWeights` method, according to the chosen optimization technique (SGD, ADAM), we update the weights using the calculated gradients on the batch_size sample. Thus, we somehow take into account the gradients calculated on the entire batch_size sample

Reply:
>we calculate the gradient and summarize it for all examples in the current batch sample.

Oh, I will look into more. Could you please point to where we summerize the loss ?

Reply:
https://github.com/Samsung/ONE/blob/ad3942d79371e27e3534f3274a913605a75a1f78/onert-micro/onert-micro/src/core/train/OMTrainingRuntimeModule.cpp#L448-L454 - here summarize calculated gradients (SGD).
https://github.com/Samsung/ONE/blob/ad3942d79371e27e3534f3274a913605a75a1f78/onert-micro/onert-micro/src/core/train/OMTrainingRuntimeModule.cpp#L471-L473 - here updated weigths